### PR TITLE
[WIP] Refactor sales provider

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         |1.7
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | based on #11405
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -209,5 +209,23 @@
 
         <service id="sylius.order_item_names_setter" class="Sylius\Component\Core\Order\OrderItemNamesSetter" />
         <service id="Sylius\Component\Core\Order\OrderItemNamesSetterInterface" alias="sylius.order_item_names_setter" />
+
+        <service id="sylius.last_year_sales_summary_by_postgres_provider" class="Sylius\Component\Core\Provider\LastYearSalesSummaryByPostgresProvider">
+            <argument type="service" id="sylius.repository.order"/>
+            <tag name="sylius.sales_provider" database-platform="postgres" />
+        </service>
+
+        <service id="sylius.last_year_sales_summary_by_mysql_provider" class="Sylius\Component\Core\Provider\LastYearSalesSummaryByMySQLProvider">
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:218]:
```suggestion
        <service id="sylius.last_year_sales_summary_by_mysql_provider" class="Sylius\Component\Core\Provider\MySQLLastYearSalesSummaryByProvider">
```

Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:218]:
```suggestion
        <service id="sylius.last_year_sales_summary.mysql_provider" class="Sylius\Component\Core\Provider\MySQLLastYearSalesSummaryByProvider">
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 2]
```diff
@@ -209,5 +209,23 @@
 
         <service id="sylius.order_item_names_setter" class="Sylius\Component\Core\Order\OrderItemNamesSetter" />
         <service id="Sylius\Component\Core\Order\OrderItemNamesSetterInterface" alias="sylius.order_item_names_setter" />
+
+        <service id="sylius.last_year_sales_summary_by_postgres_provider" class="Sylius\Component\Core\Provider\LastYearSalesSummaryByPostgresProvider">
+            <argument type="service" id="sylius.repository.order"/>
+            <tag name="sylius.sales_provider" database-platform="postgres" />
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:215]:
```suggestion
            <tag name="sylius.sales_provider" database-platform="default" />
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 3]
```diff
@@ -209,5 +209,23 @@
 
         <service id="sylius.order_item_names_setter" class="Sylius\Component\Core\Order\OrderItemNamesSetter" />
         <service id="Sylius\Component\Core\Order\OrderItemNamesSetterInterface" alias="sylius.order_item_names_setter" />
+
+        <service id="sylius.last_year_sales_summary_by_postgres_provider" class="Sylius\Component\Core\Provider\LastYearSalesSummaryByPostgresProvider">
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:213]:
```suggestion
        <service id="sylius.last_year_sales_summary.postgres_provider" class="Sylius\Component\Core\Provider\DefaultLastYearSalesSummaryByProvider">
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 4]
```diff
@@ -209,5 +209,23 @@
...
+        <service id="sylius.last_year_sales_summary_by_mysql_provider" class="Sylius\Component\Core\Provider\LastYearSalesSummaryByMySQLProvider">
+            <argument type="service" id="sylius.repository.order"/>
+            <tag name="sylius.sales_provider" database-platform="mysql" />
+        </service>
+
+        <service id="sylius.last_year_sales_provider" class="Sylius\Component\Core\Provider\LastYearSalesSummaryProvider">
+            <argument type="collection">
+                <argument key="mysql" type="service" id="sylius.last_year_sales_summary_by_mysql_provider" />
+                <argument key="postgres" type="service" id="sylius.last_year_sales_summary_by_postgres_provider" />
+            </argument>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:227]:
```suggestion
```
Ref. https://symfony.com/doc/current/service_container/service_subscribers_locators.html#indexing-the-collection-of-services

File: src/Sylius/Component/Core/Dashboard/SalesDataProvider.php [hunk 5]
```diff
@@ -13,50 +13,34 @@
...
 
 /**
  * @experimental
  */
 final class SalesDataProvider implements SalesDataProviderInterface
 {
-    /** @var EntityManagerInterface */
-    private $entityManager;
+    /** @var LastYearSalesSummaryProviderInterface */
+     private $salesSummaryProvider;
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Dashboard/SalesDataProvider.php:25]:
```suggestion
    private $salesSummaryProvider;
```

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryByMySQLProvider.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            ->select("DATE_FORMAT(o.checkoutCompletedAt, '%m.%y') AS date")
+            ->addSelect("SUM(o.total) as total")
+            ->where('o.channel', ':channel')
+            ->andWhere('o.checkoutCompletedAt BETWEEN :startDate AND :endDate')
+            ->andWhere('o.paymentState = :state')
+            ->groupBy('date')
+            ->setParameter('channel', $channel)
+            ->setParameter('startDate', $startDate)
+            ->setParameter('endDate', $endDate)
+            ->setParameter('state', OrderPaymentStates::STATE_PAID);
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryByMySQLProvider.php:44]:
```suggestion
            ->setParameter('state', OrderPaymentStates::STATE_PAID)
        ;

```

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryByPostgresProvider.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    public function __construct(OrderRepositoryInterface $orderRepository)
+    {
+        $this->orderRepository = $orderRepository;
+    }
+
+    /** @psalm-suppress PossiblyUndefinedMethod */
+    public function getSalesSummary(ChannelInterface $channel, \DateTimeInterface $startDate, \DateTimeInterface $endDate): array
+    {
+        $queryBuilder = $this->orderRepository->createQueryBuilder('o')
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryByPostgresProvider.php:34]: I would prefer to make this implementation called default and return empty data, as we are not able to test it right now (or to set up PostgreSQL build on Travis)

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryProvider.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    /** @var ServiceLocator */
+    private $locator;
+
+    public function __construct(ServiceLocator $locator)
+    {
+        $locator = $this->locator;
+    }
+
+
+
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryProvider.php:30]:
```suggestion
```

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryProvider.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    public function getSalesSummary(ChannelInterface $channel, \DateTimeInterface $startDate, \DateTimeInterface $endDate): array
+    {
+        return [];
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryProvider.php:33]:
```suggestion
        if (!$this->locator->has($connection) {
            $provider = $this->locator->get('default');
        } else {
            $provider = $this->locator->get($connection);
        }
        
        return $provider-> getSalesSummary($channel, $start, $end);
```

Where `$connection` is taken from the doctrine entity manager. (and without else)

Comment (lchrusciel) [reply src/Sylius/Component/Core/Provider/LastYearSalesSummaryProvider.php:33]: The locator logic could be extracted to separate service to make it easier to test (we should pass connection as an argument there)

File: src/Sylius/Component/Core/Dashboard/SalesDataProvider.php [hunk 10]
```diff
@@ -24,40 +26,37 @@ final class SalesDataProvider implements SalesDataProviderInterface
...
         return new SalesSummary(
-            (new \DateTime('first day of next month last year')),
-            (new \DateTime('last day of this month')),
+            $startDate,
+            $endDate,
             $data
         );
     }
+
+    private function getLastYearSalesProvider(): LastYearSalesSummaryProviderInterface
```
Comment (pamil) [inline src/Sylius/Component/Core/Dashboard/SalesDataProvider.php:54]: It feels like this private method is breaking Open/Closed principle. Let's say I want to add a dedicated `LastYearSalesSummaryProviderInterface` instance - it'd require me to overwrite any function like this in any place `LastYearSalesSummaryProviderInterface` is used in.

This should be an internal implementation detail of `LastYearSalesSummaryProviderInterface` - I shouldn't be even aware of it while using it, my application code shouldn't be concerned with infrastructural stuff.

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryProviderInterface.php [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\Provider;
+
+use Sylius\Component\Core\Dashboard\SalesSummaryInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+
+interface LastYearSalesSummaryProviderInterface
```
Comment (pamil) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryProviderInterface.php:19]: Is it really `LastYear`? It feels like I could pass any start and end dates here. I think we can go with `SalesSummaryProviderInterface`.

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryByMySQLProvider.php [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+{
+    /** @var OrderRepositoryInterface|EntityRepository */
+    private $orderRepository;
+
+    public function __construct(OrderRepositoryInterface $orderRepository)
+    {
+        $this->orderRepository = $orderRepository;
+    }
+
+    /** @psalm-suppress PossiblyUndefinedMethod */
```
Comment (pamil) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryByMySQLProvider.php:31]: Is it needed for the whole method? It'd be safer to add it to a single expression instead.

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryByMySQLProvider.php [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\Provider;
+
+use Doctrine\ORM\EntityRepository;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\OrderPaymentStates;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+
+final class LastYearSalesSummaryByMySQLProvider implements LastYearSalesSummaryProviderInterface
```
Comment (pamil) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryByMySQLProvider.php:21]: `MySqlSalesSummaryProvider`?

File: src/Sylius/Component/Core/Provider/LastYearSalesSummaryProviderInterface.php [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+{
+    public function getSalesSummary(
+        ChannelInterface $channel,
+        \DateTimeInterface $startDate,
+        \DateTimeInterface $endDate
+    ): array;
```
Comment (pamil) [inline src/Sylius/Component/Core/Provider/LastYearSalesSummaryProviderInterface.php:25]: No contract for the output - we should introduce a DTO for this (or use `@psalm-return array<int, int> Years to sales amount`?). DTO preferred though.

**Comment by AdamKasp** [comment]
after consultation with @pamil we make desition to move this refactor to `master-branch`, to `1.7`

**Pull request closed without merging**